stackdriver/config/config.go

```diff
@@ -0,0 +1,166 @@
+)
+
+// Config encapsulates the metadata required to configure a Stackdriver client.
+type Config struct {
+	// ProjectID is the stackdriver project ID to which data is uploaded.
+	// This is not necessarily the GCP project ID where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ProjectID string
+	// GcpLocation is the GCP region or zone to which data is uploaded.
+	// This is not necessarily the GCP project ID where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	GcpLocation string
+	// ClusterName is the cluster name with which the data will be associated in Stackdriver.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ClusterName string
+	// GcpSecretName is the optional GCP service account key which will be used to
+	// authenticate with Stackdriver. If not provided, Google Application Default Credentials
+	// will be used (https://cloud.google.com/docs/authentication/production).
+	GcpSecretName string
+	// GcpSecretNamespace is the Kubernetes namespace where GcpSecretName is located.
+	// The Kubernetes ServiceAccount used by the pod that is exporting data to
+	// Stackdriver should have access to Secrets in this namespace.
+	GcpSecretNamespace string
+}
+
+// NewStackdriverConfigFromConfigMap returns a Config for the given configmap
+func NewStackdriverConfigFromConfigMap(config *corev1.ConfigMap) (*Config, error) {
+	if config == nil {
+		return &Config{}, nil
+	}
+	return NewStackdriverConfigFromMap(config.Data)
+}
+
+// NewStackdriverConfigFromMap returns a Config for the given map
+func NewStackdriverConfigFromMap(config map[string]string) (*Config, error) {
+	sc := &Config{}
+
+	if pi, ok := config[ProjectIDKey]; ok {
+		sc.ProjectID = pi
+	}
+
+	if gl, ok := config[GcpLocationKey]; ok {
+		sc.GcpLocation = gl
+	}
+
+	if cn, ok := config[ClusterNameKey]; ok {
+		sc.ClusterName = cn
+	}
+
+	if gsn, ok := config[GcpSecretNameKey]; ok {
+		sc.GcpSecretName = gsn
+	}
+
+	if gsns, ok := config[GcpSecretNamespaceKey]; ok {
+		sc.GcpSecretNamespace = gsns
+	}
+
+	return sc, nil
+}
+
+var (
+	// kubeclient is the in-cluster Kubernetes kubeclient.
+	kubeclient *kubernetes.Clientset
+)
+
+// GetStackdriverSecret returns the Kubernetes Secret specified in the given config.
+func GetStackdriverSecret(config *Config) (*v1.Secret, error) {
+	ensureKubeclient()
+	ns := secretNamespaceDefault
+	if config.GcpSecretNamespace != "" {
+		ns = config.GcpSecretNamespace
+	}
+
+	return kubeclient.CoreV1().Secrets(ns).Get(config.GcpSecretName, metav1.GetOptions{})
+}
+
+// ConvertSecretToExporterOption converts a Kubernetes Secret to an OpenCensus Stackdriver Exporter Option.
+func ConvertSecretToExporterOption(secret *v1.Secret) option.ClientOption {
+	return option.WithCredentialsJSON(secret.Data[secretDataFieldKey])
+}
+
+func isKubeclientInitialized() bool {
+	mtx.RLock()
+	defer mtx.Unlock()
```

Comment:
I think this should be `.RUnlock()`?

Reply:
Yeah, good catch! I was wondering why the test was failing when I ran it with a count > 1...